### PR TITLE
Add new tap: Google Analytics. Use source code tap from https://gitla…

### PR DESCRIPTION
**Pull request by dustymaxa**
Add google-analytics-tap. 
Source code tap from https://gitlab.com/meltano/tap-google-analytics.
Note: PyPi package required.

### CODE REVIEW

**Comment by koszti** [comment]
thanks @dustymaxa for making this, this is great. I'll review this in the next few days and will keep you posted.

**Comment by dustymaxa** [comment]
Hi @koszti, any updates on this?

Thanks

**Comment by koszti** [comment]
hey @dustymaxa, we were busy in the last few weeks and had no chance to review your PR earlier.

First of all the tap is working nicely and extracting data from GA, amazing! We forked the tap and released 1.0.0 to PyPI at https://pypi.org/project/pipelinewise-tap-google-analytics/

To make it fully compatible with the PipelineWise framework can you please update the PR with these small items:
* Create `singer-connectors/contrib/tap-google-analytics\requirements.txt` and add `pipelinewise-tap-google-analytics==1.0.0` . (Please note the new contrib folder. We'd like to speed up feedback to external contributors by separating connectors to `core` and `contrib` items. We'll modify the `install.sh` according to this)
* Please exclude `install.sh` from the PR
* Please add a few sentences about authentication methods (OAuth and service accouns) to `docs/connectors/taps/google_analytics.rst`. This is well detailed in the tap `README.md` but would be nice to have something in the PipelineWise doc as well
* Please add `key_file_location` optional property into the example YAML in the documentation. Maybe you can comment it out and can leave `oauth_credentials` as default but would be nice to mention in the example YAML as well that both service account and oauth tokens are supported
* Please add new `tap_google_analytics.yml.sample` file into `https://github.com/transferwise/pipelinewise/tree/master/pipelinewise/cli/samples` with the same modified YAML from the previous bullet point. In this way the CLI will generate sample yamls for every taps automatically by `pipelinewise init --name my_datapipes`.

File: docs/connectors/taps/google_analytics.rst [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+      # Method: OAuth authentication
+      view_id: "<view-id>"
+      start_date: "2010-01-01"  # specifies the date at which the tap will begin pulling data
+
+      oauth_credentials:
+        client_id: "<client-id>"
+        client_secret: "<oauth-client-id>"               # Plain string or vault encrypted
+        access_token: "<access-token>"                   # Plain string or vault encrypted
+        refresh_token: "<refresh-token>"                 # Plain string or vault encrypted
+
```
Comment (koszti) [inline docs/connectors/taps/google_analytics.rst]: can you please add `# key_file_location: "full-path-to-client_secrets.json"` with some comments that connecting via service accounts is also supported

File: docs/connectors/taps/google_analytics.rst [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+-----------
+
+
+Configuring what to replicate
+'''''''''''''''''''''''''''''
+
+PipelineWise configures every tap with a common structured YAML file format.
+A sample YAML for Google Analytic replication can be generated into a project directory by
+following the steps in the :ref:`generating_pipelines` section.
+
```
Comment (koszti) [inline docs/connectors/taps/google_analytics.rst]: Can you please add a few sentences about authentication methods (OAuth and service accouns) to docs/connectors/taps/google_analytics.rst. This is well detailed in the tap README.md but would be nice to have something in the PipelineWise doc as well

Also, can you please mention that STATE messages and incremental replication is currently not supported by this tap? This is quite important and documented int the tap README.md but would be nice to clarify this in the pipelinewise documentation at well.

**Comment by dustymaxa** [comment]
@koszti thanks for your feedback!

Your comments have been applied. Tell me if anything is still missing.

Dusty

**Comment by koszti** [comment]
thanks for this @dustymaxa , this looks good now. Just wanted to give you a heads up that unfortunately we don't merge PRs at this time of the year so still need to wait a few more days but will do it in the beginning of the new year.

**Comment by dustymaxa** [comment]
Hi @koszti, 

I have updated the pull request. I have sync this fork repository with your upstream repository. 
FYI the install.sh script crashes when installing the tap-google-analytics connector. Do you still have to update the install.sh for this? 
In previous comment you said : 

> (Please note the new contrib folder. We'd like to speed up feedback to external contributors by separating connectors to core and contrib items. **We'll modify the install.sh according to this**)


```
--------------------------------------------------------------------------
Installing tap-google-analytics connector...
--------------------------------------------------------------------------
ERROR: Directory not exists and does not look like a valid singer connector: /app/singer-connectors/tap-google-analytics
The command '/bin/sh -c cd /app     && ./install.sh --connectors=all --acceptlicenses --nousage --notestextras     && ln -s /root/.pipelinewise /app/.pipelinewise' returned a non-zero code: 1
```

Thank you,

Dusty

**Comment by Jude188** [comment]
Is this the best way to go? The WIP official Google Analytics tap already supports state (quite cleverly). Would it make more sense to wait and use that one?

**Comment by gouline** [comment]
Is there anything I can help with to get this merged? GA is the only tap I'm missing from PipelineWise, so happy to help if there's anything left to do.

**Comment by koszti** [comment]
@gouline this PR has not been merged for no any particular reasons.

How do we feel about the forked meltano tap-google-analytics that this tap is based on?
Do we need to update the fork from the upstream?
PPW forked: https://github.com/transferwise/pipelinewise-tap-google-analytics/tree/master/tap_google_analytics
Meltano upstream: https://gitlab.com/meltano/tap-google-analytics/-/tree/master/

Like @jude188 mentioned, singer in the meantime made a tap-google-analytics as well (https://github.com/singer-io/tap-google-analytics) and wondering if we should use the meltano or the singer implementation of the tap?

Like @jude188 mentioned the singer tap-GA supports states and hence the incremental replication, which sounds like a big improvement. Currently this tap in the PR doesn't support incremental method.

What do you think, which direction we should go to with this tap?

**Comment by gouline** [comment]
Thanks for getting back to me so quickly.

Happy to go either way, slightly leaning towards sticking with the Meltano one at least for now. Reason being the chunkier documentation they have going on. Also the official one explicitly calls out that it's under development and incomplete, so possibly more breaking changes.

Let's see if @dustymaxa is still interested to sync their fork against upstream, otherwise I can create my own one (I'd rather give them the credit for their work).

As for the tap, we could start with what's currently in https://github.com/transferwise/pipelinewise-tap-google-analytics and I can create a PR with latest changes in the GitLab upstream.

**Comment by dustymaxa** [comment]
Hi guys,

sure @gouline go head. Both solutions are good for me too. The singer implementing might sound like a better idea. 
Keep me posted when the GA tap will be supported with pipelinewise :)!

**Comment by gouline** [comment]
@dustymaxa Would you mind syncing your fork against upstream (since this PR is from your fork)? Or would you rather I created a separate one?

**Comment by dustymaxa** [comment]
@gouline I don't have much time to work on this pr right now. So you should maybe create a separate one

**Comment by koszti** [comment]
@gouline 1.1.0 pipelinewise-tap-google-analytics is now available in PyPI at https://pypi.org/project/pipelinewise-tap-google-analytics/

Please update [requirements.txt](https://github.com/transferwise/pipelinewise/pull/267/files#diff-3425260ba55b3ec3fbaf7be29a5353c6) accordingly and fix the conflicts. Should be straightforward.

**Comment by gouline** [comment]
Done. Refer to #426 for updated changes.

**Comment by koszti** [comment]
duplicate of #426. Cleansed PR available in #426

**Pull request closed without merging**